test/Transition-test.js

```diff
@@ -131,6 +131,66 @@ describe('Transition', () => {
     inst.setProps({ in: true })
   })
 
+  describe('render prop `done` callback', () => {
+    it('should end entering transition when child triggers `done`', done => {
+      class DoneTrigger extends React.Component {
+        componentDidUpdate() {
+          if (this.props.status === ENTERING) this.props.done()
+          if (this.props.status === ENTERED) done()
+        }
+
+        render() {
+          return <div />
+        }
+      }
+
+      mount(
+        <Transition
+          in
+          appear
+          timeout={10000}
+          mountOnEnter
+          unmountOnExit
+        >
+          {(status, _childProps, doneCB) => (
+            <DoneTrigger status={status} done={doneCB} />
+          )}
+        </Transition>
+      )
+    })
+
+    it('should end `exiting` transition when child triggers `done`', done => {
+      class DoneTrigger extends React.Component {
+        componentDidMount() {
+          expect(this.props.status).toEqual(ENTERED)
+        }
+
+        componentDidUpdate() {
+          if (this.props.status === EXITING) this.props.done()
+          if (this.props.status === EXITED) done()
+        }
+
+        render() {
+          return <div />
+        }
+      }
+
+      const wrapper = mount(
+        <Transition
+          in
+          timeout={10000}
+          mountOnEnter
+        >
+          {(status, _childProps, doneCB) => (
+            <DoneTrigger status={status} done={doneCB} />
+          )}
+        </Transition>
+      )
+
+      wrapper.setProps({in: false})
+    })
+  })
+
```

Comment:
we needs tests, for when components aren't transitioning. Should also make sure that calling done multiple times in a row doesn't break anything.

Reply:
I didn’t want to offend you and do admit that you understand the context better then I. But looking at the comments in code about safeSetState (like “it shouldn’t be needed, but somehow it avoids the problems, so let’s use that for now”), I thought it could be one of long standing todos to address. After spending quite a lot of time reading the code, I’ve noticed that the library uses the old context API which is about to be abandoned, has no consistent code style (did you consider prettier?). I thought it’s one of these issues I could try to help with. If you believe the current safeSetState solution is optimal and is required indeed, maybe the comment that it “shouldn’t be needed” should be removed not to confuse people who have less context but still want to contribute.

Regarding the PR, it’s still a WIP, I’ll ping you when it will be ready for review.

Sure I scratch my own itch in the first place, but I’d be happy to improve the library for everybody along the way, and I do appreciate your help with my efforts on both.